core/server/master/src/main/java/alluxio/master/MasterProcess.java

```diff
@@ -92,4 +94,29 @@ private Factory() {} // prevent instantiation
    * @return configuration information list
    */
   List<ConfigProperty> getConfiguration();
+
+  /**
+   * Returns a master id for the given master, creating one if the master is new.
+   *
+   * @param hostname the master hostname
+   * @return the master id for this master
+   */
+  long getMasterId(String hostname);
+
+  /**
+   * Updates metadata when a standby master periodically heartbeats with the leader master.
+   *
+   * @param masterId the master id
+   * @return whether the master should re-register
+   */
+  boolean masterHeartbeat(long masterId);
+
+  /**
+   * Updates metadata when a standby master registers with the leader master.
+   *
+   * @param masterId the master id of the standby master registering
+   * @param options the options that contains master configuration
+   * @throws NoMasterException if masterId cannot be found
```

Comment:
why does `masterHeartbeat` use `boolean` to indicate that the master ID wasn't found, but `masterRegister` uses `NoMasterException`?

Reply:
In the metaMasterSync which manages the master communications, I use the logic that if master id is not set, the masterHeartbeat will ask the master to re-register. When we register, we always get the master Id first and then call masterRegister. If everything goes as expected, the masterId should be found and will not trigger the NoMasterException.
```
MetaMasterSync.heartbeat()
      boolean shouldReRegister = mMasterClient.heartbeat(mMasterId.get());
      if (shouldReRegister) {
        mMasterId.set(mMasterClient.getId(mMasterHostname));
        registerWithMaster();
      }
```

So that masterId is not found is reasonable and expected in masterHeartbeat and not reasonable and expected in masterRegister.

Reply:
Got it, using exception for one but not the other SGTM